Add type annotations to core.py

xanderdunn opened the pull request:
I tried my hand at adding type annotations for #1210. A few things to note here:

Adding return type annotations to JIT compiled functions in core.py sometimes broke the JIT compilation because it introduced a new element for the return value annotation called `return`, which is not in the args list. To solve this, I filtered out this annotation in the list, see the change in the jit.py file. 

Additionally, I found that adding certain type annotations to arguments would cause a runtime error. For example, modifying the `def zeros_like(input)` function declaration in core.py to `def zeros_like(input: tensor)` causes this runtime error:
```
Traceback (most recent call last):
  File "./src/fused_gelu.py", line 4, in <module>
    import triton
  File "/home/xander/dev/triton/python/triton/__init__.py", line 29, in <module>
    from . import language
  File "/home/xander/dev/triton/python/triton/language/__init__.py", line 8, in <module>
    from . import libdevice
  File "/home/xander/dev/triton/python/triton/language/libdevice.py", line 4, in <module>
    from . import core, extern
  File "/home/xander/dev/triton/python/triton/language/core.py", line 1294, in <module>
    def zeros_like(input: tensor) -> tensor:
  File "/home/xander/dev/triton/python/triton/runtime/jit.py", line 421, in jit
    return decorator(fn)
  File "/home/xander/dev/triton/python/triton/runtime/jit.py", line 414, in decorator
    return JITFunction(
  File "/home/xander/dev/triton/python/triton/runtime/jit.py", line 323, in __init__
    self.run = self._make_launcher()
  File "/home/xander/dev/triton/python/triton/runtime/jit.py", line 291, in _make_launcher
    exec(src, scope)
  File "<string>", line 3
    sig_key =  ,
               ^
SyntaxError: invalid syntax
```
Some issue with the JIT compilation that I haven't figured out yet. 

These code changes compiled and ran fine for me on my [split gelu examples](https://github.com/xanderdunn/fused-gelu-testing/blob/main/src/fused_gelu.py), as well as the 03_ matrix multiplication and 01_ vector addition tutorial files.

ptillet commented on python/triton/language/core.py:
```diff
@@ -887,7 +888,7 @@ def store(pointer, value, mask=None, cache_modifier="", eviction_policy="", _bui
     # value can be constexpr
     value = _to_tensor(value, _builder)
-    if _constexpr_to_value(mask) is not None:
+    if mask is not None and _constexpr_to_value(mask) is not None:
```
I don't really understand the point of this additional check, since `_constexpr_to_value` returns `None` when its input is `None`

xanderdunn replied:
That's correct, this has no impact on functionality, but it silences a type checker warning that `_to_tensor` does not take `None`. The type checker does not evaluate `_constexpr_to_value` to determine this. Making it explicit silences the warning.

ptillet replied:
Couldn't we silence the typechecker with some return type hint on `_constexpr_to_value` ? In general I'm pretty hostile to creating redundancy in the code for the sake of silencing the typechecker.

ptillet commented:
I think a lot of these type annotations are actually incomplete and don't really account for the implicit conversions that may happen

xanderdunn commented:
@ptillet thanks for taking a look. Definitely agree that there's more type work to be done here. Can you give an example of an implicit conversion that isn't being handled well? 

For elements that are fed to `_to_tensor` I've created the type alias `Tensorable = Union[bool, float, int, tensor, constexpr]` which takes into account the many different types that can be converted to `tl.tensor`. However, there are other situations where the input is not passed to `_to_tensor`. For example, the `load()` function directly passes the `pointer` argument to `semantic.load`, which shows in its type declaration that it's expecting a `tl.tensor`. So in this case I believe it's correct to annotate `load(pointer: tensor` directly, no conversion is expected?

xanderdunn commented on python/triton/language/core.py:
```diff
@@ -1277,7 +1278,7 @@ def swizzle2d(i, j, size_i, size_j, size_g):
 
 @triton.jit
-def zeros(shape, dtype):
+def zeros(shape, dtype: dtype) -> tensor:
```
@ptillet The `dtype: dtype` annotation here may be an incorrect assumption. I see in `full` it does a potential `constexpr` conversion, so this is not necessarily a `dtype`. There may be a few others, but I believe I largely avoided this. I'll fix this one tomorrow.

xanderdunn commented on python/triton/language/core.py:
```diff
@@ -843,7 +844,7 @@ def dot(input, other, allow_tf32=True, _builder=None):
 
 @builtin
-def load(pointer, mask=None, other=None, cache_modifier="", eviction_policy="", volatile=False, _builder=None):
+def load(pointer: tensor, mask: Optional[tensor]=None, other: Optional[tensor]=None, cache_modifier="", eviction_policy="", volatile=False, _builder=None) -> tensor:
```
`mask` and `other` here are  passed to `_to_tensor` so I believe their types should instead be `Optional[Tensorable]`

xanderdunn commented:
The latest commits resolve the implicit type conversion errors that I was able to find. 

This error in the original post:
```
...
  File "/home/xander/dev/triton/python/triton/runtime/jit.py", line 291, in _make_launcher
    exec(src, scope)
  File "<string>", line 3
    sig_key =  ,
               ^
SyntaxError: invalid syntax
```
is happening because some non-constexpr arguments are now being included in `self.constexprs` in the `JITFunction` `__init__` in jit.py. Adding a type annotation to a function argument causes the argument to be captured here:
```python
self.constexprs = [self.arg_names.index(ann) for ann in self.__annotations__.keys()]
```

For example, adding the type annotation to `def zeros_like(input: tensor)` causes the `input` argument to show up in the `self.constexprs` list, which is not right. This needs to be fixed for this pull request to be mergeable. Otherwise we may be capturing type annotated args as constexprs when they are not. I have not yet figure out how to properly capture which ones are constexprs.

ptillet commented:
Yeah, this PR seems like a minor but net improvement to the codebase -- provided that we can find a proper fix for the typechecker warning mentioned in the comments. I also want to take some time to think of a better name than `Tensorable`

xanderdunn commented:
Ok, I conflated two separate issues:
1. Removing type warnings within the triton code base
2. Adding type annotations for the convenience of projects making use of triton as a library

The first one is not the goal of this PR, the second one is. So, I've removed the 3 lines of changes I made like `if mask is not None` to silence the type checker within triton. I think that's out of scope for this PR. 

I also renamed `Tensorable` to `TensorConvertible`. This is more self-explanatory but it's also a bit verbose. I'm open to other name suggestions as well.

lezcano commented:
In PyTorch we use `TensorLike` :)

xanderdunn commented:
`TensorLike` is definitely better @lezcano, I just changed it. Thanks!

ptillet commented:
I like TensorLike :) I'll merge this

ptillet commented:
sorry, I dropped the ball on that and now there are some conflicts. Would you mind resolving the conflicts?
